execution/executor/src/speculation_cache/mod.rs

```diff
@@ -0,0 +1,244 @@
+    transactions: Vec<Transaction>,
+    // The pointers to all the children blocks.
+    children: Vec<Arc<Mutex<SpeculationBlock>>>,
+    // The speculative execution result.
+    output: ProcessedVMOutput,
+    // A pointer to the global block map keyed by id to achieve O(1) lookup time complexity.
+    block_map: Arc<Mutex<HashMap<HashValue, Weak<Mutex<SpeculationBlock>>>>>,
+}
+
+impl SpeculationBlock {
+    pub(crate) fn new(
+        id: HashValue,
+        transactions: Vec<Transaction>,
+        output: ProcessedVMOutput,
+        block_map: Arc<Mutex<HashMap<HashValue, Weak<Mutex<SpeculationBlock>>>>>,
+    ) -> Self {
+        Self {
+            id,
+            transactions,
+            children: vec![],
+            output,
+            block_map,
+        }
+    }
+
+    pub(crate) fn id(&self) -> HashValue {
+        self.id
+    }
+
+    pub(crate) fn transactions(&self) -> &Vec<Transaction> {
+        &self.transactions
+    }
+
+    pub(crate) fn add_child(&mut self, child: Arc<Mutex<SpeculationBlock>>) {
+        self.children.push(child)
+    }
+
+    pub(crate) fn output(&self) -> &ProcessedVMOutput {
+        &self.output
+    }
+}
+
+/// drop() will clean the current block entry from the global map.
+impl Drop for SpeculationBlock {
+    fn drop(&mut self) {
+        self.block_map
+            .lock()
+            .unwrap()
+            .remove(&self.id())
+            .expect("Speculation block must exist in block_map before being dropped.");
+        debug!("Speculation block {} is dropped.", self.id())
+    }
+}
+
+/// SpeculationCache implements the block tree structrue. The tree is reprensented by a root block id,
+/// all the children of root and a global block map. Each block is an Arc<Mutx<SpeculationBlock>>
+/// with ref_count = 1. For the chidren of the root, the sole owner is `heads`. For the rest, the sole
+/// owner is their parent block. So when a block is dropped, all its descendants will be dropped
+/// recursively. In the meanwhile, wheir entries in the block map will be removed by each block's drop().
+pub(crate) struct SpeculationCache {
+    synced_trees: ExecutedTrees,
+    committed_trees: ExecutedTrees,
+    // The id of root block.
+    committed_block_id: HashValue,
+    // The chidren of root block.
+    heads: Vec<Arc<Mutex<SpeculationBlock>>>,
+    // A pointer to the global block map keyed by id to achieve O(1) lookup time complexity.
+    // It is optional but an optimization.
+    block_map: Arc<Mutex<HashMap<HashValue, Weak<Mutex<SpeculationBlock>>>>>,
+}
+
+impl SpeculationCache {
+    pub fn new() -> Self {
+        Self {
+            synced_trees: ExecutedTrees::new_empty(),
+            committed_trees: ExecutedTrees::new_empty(),
+            heads: vec![],
+            block_map: Arc::new(Mutex::new(HashMap::new())),
+            committed_block_id: *PRE_GENESIS_BLOCK_ID,
+        }
+    }
+
+    pub(crate) fn new_with_startup_info(startup_info: StartupInfo) -> Self {
```

Comment:
these pub(crate) are meaningless if you define you struct as pub(crate)

Reply:
I am not sure what's the convention here.